File: Ch09/PerformanceTuning/PerformanceTuning/ViewController.m
//
//  ViewController.m
//  PerformanceTuning
//
//  Created by Rob Warner on 11/3/14.
//  Copyright (c) 2014 Michael Privat and Rob Warner. All rights reserved.
//

#import "ViewController.h"
#import "FetchAll.h"
#import "AppDelegate.h"
#import "Persistence.h"
#import "DidFault.h"
#import "SingleFault.h"
#import "BatchFault.h"
#import "Prefetch.h"
#import "Cache.h"
#import "Uniquing.h"
#import "Predicate.h"
#import "Subquery.h"
#import "BatchUpdate.h"

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  self.startTime.text = @"";
  self.stopTime.text = @"";
  self.elapsedTime.text = @"";
  self.results.text = @"";
  
  self.tests = @[[[FetchAll alloc] init],
                 [[DidFault alloc] init],
                 [[SingleFault alloc] init],
                 [[BatchFault alloc] init],
                 [[Prefetch alloc] init],
                 [[Cache alloc] init],
                 [[Uniquing alloc] init],
                 [[Predicate alloc] init],
                 [[Subquery alloc] init],
                 [[BatchUpdate alloc] init]];
}

#pragma mark - UIPickerViewDataSource methods

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  return [self.tests count];
}

#pragma mark - UIPickerViewDelegate methods

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
  id <PerfTest> test = self.tests[row];
  return [[test class] description];
}

#pragma mark - Handlers

- (IBAction)runTest:(id)sender {
  // Get the selected test
  id <PerfTest> test = self.tests[[self.testPicker selectedRowInComponent:0]];
  AppDelegate *delegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
  NSManagedObjectContext *context = delegate.persistence.managedObjectContext;
  
  // Clear out any objects so we get clean test results
  [context reset];
  
  // Mark the start time, run the test, and mark the stop time
  NSDate *start = [NSDate date];
  NSString *results = [test runTestWithContext:delegate.persistence.managedObjectContext];
  NSDate *stop = [NSDate date];
  
  // Update the UI with the test results
  self.startTime.text = [start description];
  self.stopTime.text = [stop description];
  self.elapsedTime.text = [NSString stringWithFormat:@"%.03f seconds", [stop timeIntervalSinceDate:start]];
  self.results.text = results;
}

@end

File: Ch09/PerformanceTuningSwift/PerformanceTuningSwift/ViewController.swift
//
// From the book Pro iOS Persistence
// Michael Privat and Rob Warner
// Published by Apress, 2014
// Source released under The MIT License
// http://opensource.org/licenses/MIT
//
// Contact information:
// Michael: @michaelprivat -- http://michaelprivat.com -- mprivat@mac.com
// Rob: @hoop33 -- http://grailbox.com -- rwarner@grailbox.com
//

import UIKit

class ViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource {

    @IBOutlet weak var startTime: UILabel!
    @IBOutlet weak var stopTime: UILabel!
    @IBOutlet weak var elapsedTime: UILabel!
    @IBOutlet weak var results: UITextView!
    @IBOutlet weak var testPicker: UIPickerView!
    
    var tests = [PerfTest]()
    
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return self.tests.count
    }
    
    func pickerView(pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String! {
        let test = self.tests[row]
        let fullName = _stdlib_getDemangledTypeName(test)
        let tokens = split(fullName, { $0 == "." })
        return tokens.last
    }
    
    @IBAction func runTest(sender: AnyObject) {
        // Get the selected test
        let test = self.tests[self.testPicker.selectedRowInComponent(0)]
        
        let appDelegate = UIApplication.sharedApplication().delegate as AppDelegate
        let context = appDelegate.persistence?.managedObjectContext
        if let context = context {
            // Clear out any objects so we get clean test results
            context.reset()
            
            // Mark the start time, run the test, and mark the stop time
            let start = NSDate()
            let testResults = test.runTestWithContext(context)
            let stop = NSDate()
            
            let formatter = NSDateFormatter()
            formatter.timeStyle = .MediumStyle
            
            // Update the UI witht the test results
            self.startTime.text = formatter.stringFromDate(start)
            self.stopTime.text = formatter.stringFromDate(stop)
            self.elapsedTime.text = NSString(format: "%.03f seconds", stop.timeIntervalSinceDate(start))
            self.results.text = testResults
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
      
        self.startTime.text = ""
        self.stopTime.text = ""
        self.elapsedTime.text = ""
        self.results.text = ""
        
        self.tests.append(FetchAll())
        self.tests.append(DidFault())
        self.tests.append(SingleFault())
        self.tests.append(BatchFault())
        self.tests.append(Prefetch())
        self.tests.append(Cache())
        self.tests.append(Uniquing())
        self.tests.append(Predicate())
        self.tests.append(Subquery())
        self.tests.append(BatchUpdate())
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}

